src/System.Net.Http/src/System/Net/Http/DiagnosticsHandler.cs

```diff
@@ -28,36 +30,62 @@ internal static bool IsEnabled()
             return s_diagnosticListener.IsEnabled();
         }
 
-        protected internal override Task<HttpResponseMessage> SendAsync(HttpRequestMessage request,
+        protected internal override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request,
             CancellationToken cancellationToken)
         {
             //HttpClientHandler is responsible to call DiagnosticsHandler.IsEnabled() before forwarding request here.
             //This code will not be reached if no one listens to 'HttpHandlerDiagnosticListener', unless consumer unsubscribes
             //from DiagnosticListener right after the check. So some requests happening right after subscription starts
             //might not be instrumented. Similarly, when consumer unsubscribes, extra requests might be instumented
 
-            Guid loggingRequestId = Guid.NewGuid();
-            LogHttpRequest(request,  loggingRequestId);
-
-            Task<HttpResponseMessage> responseTask = base.SendAsync(request, cancellationToken);
-
-            LogHttpResponse(responseTask, loggingRequestId);
-            return responseTask;
-        }
-
-        #region private
+            Activity activity = null;
+            Guid loggingRequestId = Guid.Empty;
+            
+            //cache IsEnabled result for Activity name
+            if (!s_activityEventIsChecked)
```

Comment:
I would strongly prefer to not have static caching of enabled state.   They are subtle bug farms, especially if there is contract in between (like there is here).    What if one subscriber just wanted exceptions, and then a bit later, another subscriber joins and wants Activities.   As the code is written the second subscriber does not get his events (because of cached enabled state when the first subscriber enabled the diagnosticSource).  

I have sent you a counter-proposal in E-mail.